src/Catty/Extension&Delegate&Protocol/Extensions/ViewController+Extension/ProjectDetailStoreViewControllerExtension.swift

```diff
@@ -0,0 +1,92 @@
+/**
+ *  Copyright (C) 2010-2020 The Catrobat Team
+ *  (http://developer.catrobat.org/credits)
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  An additional term exception under section 7 of the GNU Affero
+ *  General Public License, version 3, is available at
+ *  (http://developer.catrobat.org/license_additional_term)
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+import Foundation
+
+@objc
+extension ProjectDetailStoreViewController {
+
+    func download(name: String) {
+        self.storeProjectDownloader.download(
+            projectId: self.project.projectID,
+            completion: { data, storeProjectDownloaderError in
+                if let error = storeProjectDownloaderError {
+                    switch error {
+                    case .unexpectedError, .timeout:
+                        Util.defaultAlertForNetworkError()
+                    case .parse(error: _), .request(error: _, statusCode: _):
+                        Util.alert(withText: kLocalizedInvalidZip)
+                    }
+                    self.setBackDownloadStatus()
+                    return
+                }
+
+                if let cbFileManager = CBFileManager.shared() {
+                    if let url = URL(string: self.project.downloadUrl),
+                        let projectName = self.project.projectName,
+                        let projectId = self.project.projectID {
+                        debugPrint("Starting downloading project '%@' with id %@ from url: %@", projectName, projectId, url.absoluteString)
+                        cbFileManager.delegate = self
+
+                        let downloadTask = self.storeProjectDownloader.session.downloadTask(with: url)
+
+                        cbFileManager.projectTaskDict.setObject(url, forKey: downloadTask)
+                        cbFileManager.projectNameDict.setObject(projectName, forKey: downloadTask)
+                        cbFileManager.projectIDDict.setObject(projectId, forKey: downloadTask)
+                        cbFileManager.storeDownloadedProject(data, andTask: downloadTask)
```

Comment:
Since the project needs to be stored on the device almost always when the download method is called, we need this code at several places. My suggestion would be to move it into the download method for the `StoreProjectDownloader`. Maybe you can also get rid of those three dictionaries?